ulauncher/main.py

```diff
@@ -1,54 +1,68 @@
 import sys
+import argparse
 import signal
 import logging
 from functools import partial
 # This xinit import must happen before any GUI libraries are initialized.
 # pylint: disable=wrong-import-position,wrong-import-order,ungrouped-imports,unused-import
 import ulauncher.utils.xinit  # noqa: F401
-
 import gi
 gi.require_version('Gtk', '3.0')
-gi.require_version('GLib', '2.0')
 # pylint: disable=wrong-import-position
-from gi.repository import Gtk, GLib
-import dbus
-import dbus.service
-from dbus.mainloop.glib import DBusGMainLoop
-
+from gi.repository import Gio, GLib, Gtk
 from ulauncher.config import API_VERSION, VERSION, get_options
 from ulauncher.utils.wayland import is_wayland, is_wayland_compatibility_on
-from ulauncher.ui.windows.UlauncherWindow import UlauncherWindow
-from ulauncher.ui.AppIndicator import AppIndicator
 from ulauncher.utils.Settings import Settings
 from ulauncher.utils.setup_logging import setup_logging
 
-logger = logging.getLogger('ulauncher')
 
-DBUS_SERVICE = 'net.launchpad.ulauncher'
-DBUS_PATH = '/net/launchpad/ulauncher'
+class UlauncherApp(Gtk.Application):
+    # Gtk.Applications check if the app is already registered and if so,
+    # new instances sends the signals to the registered one
+    # So all methods except __init__ runs on the main app
+    window = None
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(
+            *args,
+            application_id="net.launchpad.ulauncher",
+            flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE,
+            **kwargs
+        )
+        self.connect("startup", self.setup)  # runs only once on the main instance
 
-class UlauncherDbusService(dbus.service.Object):
-    def __init__(self, window):
-        self.window = window
-        bus_name = dbus.service.BusName(DBUS_SERVICE, bus=dbus.SessionBus())
-        super().__init__(bus_name, DBUS_PATH)
+    def setup(self, _):
+        self.hold()  # Keep the app running even without a window
+        # These modules are very heavy, so we don't want to load them in the remote
+        # pylint: disable=import-outside-toplevel
+        from ulauncher.ui.windows.UlauncherWindow import UlauncherWindow
+        from ulauncher.ui.AppIndicator import AppIndicator
+        self.window = UlauncherWindow.get_instance()
+        if Settings.get_instance().get_property('show-indicator-icon'):
+            AppIndicator.get_instance().show()
 
-    @dbus.service.method(DBUS_SERVICE)
-    def toggle_window(self):
-        self.window.toggle_window()
+    def do_activate(self, *args, **kwargs):
+        self.window.show_window()
 
+    def do_command_line(self, *args, **kwargs):
```

Comment:
Should this message signature just be `def do_command_line(self, appcl):` or similar? Looks like according to doc that is all that will be passed in.

Reply:
That's what I did originally, but then I get these linting errors locally (my pylint is way more recent than the docker image one though):
>ulauncher/main.py:47:4: W0221: Number of parameters was 3 in 'Application.do_command_line' and is now 2 in overridden 'UlauncherApp.do_command_line' method (arguments-differ)
>ulauncher/main.py:47:4: W0221: Variadics removed in overridden 'UlauncherApp.do_command_line' method (arguments-differ)

Same for `do_activate()`.